File: src/dataset.c
//
// Created by thiagorigonatti on 29/04/26.
//

#include "dataset.h"

static int read_exact(FILE *f, void *ptr, size_t len) {
    return fread(ptr, 1, len, f) == len ? 0 : -1;
}

dataset_t g_dataset;

int dataset_load_index(const char *path) {
    FILE *f = fopen(path, "rb");
    if (!f) {
        perror(path);
        return -1;
    }

    char magic[4];
    uint32_t n = 0, k = 0, d = 0, stride = 0;
    float scale = 0.0f;
    if (read_exact(f, magic, 4) != 0 || memcmp(magic, "IVF6", 4) != 0 ||
        read_exact(f, &n, sizeof(n)) != 0 ||
        read_exact(f, &k, sizeof(k)) != 0 ||
        read_exact(f, &d, sizeof(d)) != 0 ||
        read_exact(f, &stride, sizeof(stride)) != 0 ||
        read_exact(f, &scale, sizeof(scale)) != 0) {
        fprintf(stderr, "index invalido ou magic diferente de IVF6: %s\n", path);
        fclose(f);
        return -1;
    }

    if (k != IVF_CLUSTERS || d != DIM || stride != DIM) {
        fprintf(stderr, "index incompatível: k=%u d=%u stride=%u esperado k=%d d=%d stride=%d\n", k, d, stride, IVF_CLUSTERS, DIM, DIM);
        fclose(f);
        return -1;
    }
    if (fabsf(scale - FIX_SCALE) > 0.01f) {
        fprintf(stderr, "index incompatível: scale=%.1f esperado=%.1f\n", scale, FIX_SCALE);
        fclose(f);
        return -1;
    }

    memset(&g_dataset, 0, sizeof(g_dataset));
    g_dataset.n = (int)n;
    g_dataset.centroids = (float *)xaligned_alloc(CACHELINE, (size_t)IVF_CLUSTERS * DIM * sizeof(float));
    g_dataset.bbox_min = (int16_t *)xaligned_alloc(CACHELINE, (size_t)IVF_CLUSTERS * DIM * sizeof(int16_t));
    g_dataset.bbox_max = (int16_t *)xaligned_alloc(CACHELINE, (size_t)IVF_CLUSTERS * DIM * sizeof(int16_t));
    if (!g_dataset.centroids || !g_dataset.bbox_min || !g_dataset.bbox_max) { fclose(f); return -1; }

    if (read_exact(f, g_dataset.centroids, (size_t)IVF_CLUSTERS * DIM * sizeof(float)) != 0 ||
        read_exact(f, g_dataset.bbox_min, (size_t)IVF_CLUSTERS * DIM * sizeof(int16_t)) != 0 ||
        read_exact(f, g_dataset.bbox_max, (size_t)IVF_CLUSTERS * DIM * sizeof(int16_t)) != 0) {
        fclose(f); return -1;
    }

    uint32_t *offsets = (uint32_t *)malloc((size_t)(IVF_CLUSTERS + 1) * sizeof(uint32_t));
    if (!offsets) { fclose(f); return -1; }
    if (read_exact(f, offsets, (size_t)(IVF_CLUSTERS + 1) * sizeof(uint32_t)) != 0) {
        free(offsets); fclose(f); return -1;
    }
    for (int c = 0; c < IVF_CLUSTERS; c++) {
        g_dataset.cluster_start[c] = (int)offsets[c];
        g_dataset.cluster_end[c] = (int)offsets[c + 1];
    }
    free(offsets);

    for (int j = 0; j < DIM; j++) {
        g_dataset.dim[j] = (int16_t *)xaligned_alloc(CACHELINE, (size_t)n * sizeof(int16_t));
        if (!g_dataset.dim[j]) { fclose(f); return -1; }
    }

    const size_t CHUNK = 16384;
    int16_t *tmp = (int16_t *)malloc(CHUNK * DIM * sizeof(int16_t));
    if (!tmp) { fclose(f); return -1; }
    uint32_t done = 0;
    while (done < n) {
        uint32_t take = n - done;
        if (take > CHUNK) take = CHUNK;
        if (read_exact(f, tmp, (size_t)take * DIM * sizeof(int16_t)) != 0) {
            free(tmp); fclose(f); return -1;
        }
        for (uint32_t i = 0; i < take; i++) {
            for (int j = 0; j < DIM; j++) {
                g_dataset.dim[j][done + i] = tmp[(size_t)i * DIM + j];
            }
        }
        done += take;
    }
    free(tmp);

    g_dataset.labels = (uint8_t *)xaligned_alloc(CACHELINE, (size_t)n * sizeof(uint8_t));
    if (!g_dataset.labels) { fclose(f); return -1; }
    if (read_exact(f, g_dataset.labels, (size_t)n) != 0) { fclose(f); return -1; }

    g_dataset.orig_ids = (uint32_t *)xaligned_alloc(CACHELINE, (size_t)n * sizeof(uint32_t));
    if (!g_dataset.orig_ids) { fclose(f); return -1; }
    if (read_exact(f, g_dataset.orig_ids, (size_t)n * sizeof(uint32_t)) != 0) { fclose(f); return -1; }

    fclose(f);

    double mb = ((double)n * DIM * sizeof(int16_t) + (double)n + (double)n * sizeof(uint32_t) +
                 (double)IVF_CLUSTERS * DIM * (sizeof(float) + 2.0 * sizeof(int16_t))) / (1024.0 * 1024.0);
    fprintf(stderr, "index IVF6 carregado: N=%u K=%u scale=%.1f memoria=%.2f MB\n", n, k, scale, mb);
    return 0;
}

void dataset_free(void) {
    for (int j = 0; j < DIM; j++) free(g_dataset.dim[j]);
    free(g_dataset.labels);
    free(g_dataset.orig_ids);
    free(g_dataset.centroids);
    free(g_dataset.bbox_min);
    free(g_dataset.bbox_max);
    memset(&g_dataset, 0, sizeof(g_dataset));
}

File: zig/src/dataset.zig
const std = @import("std");
const c = @cImport({
    @cInclude("stdio.h");
});
const config = @import("config.zig");

const DIM = config.DIM;
const IVF_CLUSTERS = config.IVF_CLUSTERS;
const BLOCK_STRIDE = config.BLOCK_STRIDE;

const CFile = struct {
    fp: ?*c.FILE,

    fn open(path: [*:0]const u8) !CFile {
        const fp = c.fopen(path, "rb");
        if (fp == null) return error.FileNotFound;
        return CFile{ .fp = fp };
    }

    fn close(self: *CFile) void {
        _ = c.fclose(self.fp);
        self.fp = null;
    }

    fn read(self: *CFile, buf: []u8) !usize {
        const n = c.fread(buf.ptr, 1, buf.len, self.fp);
        if (n != buf.len and c.feof(self.fp) == 0) return error.ReadError;
        return n;
    }

    fn readInt(self: *CFile, comptime T: type) !T {
        var buf: [@sizeOf(T)]u8 = undefined;
        _ = try self.read(&buf);
        return std.mem.readInt(T, &buf, .little);
    }
};

pub const Dataset = struct {
    n: usize,
    total_blocks: usize,
    padded_n: usize,
    centroids_t: []f32,
    block_offsets: [IVF_CLUSTERS + 1]u32,
    labels: []u8,
    blocks: []i16,
    allocator: std.mem.Allocator,

    pub fn load(allocator: std.mem.Allocator, path: []const u8) !Dataset {
        const path_z = try allocator.dupeZ(u8, path);
        defer allocator.free(path_z);
        var file = try CFile.open(path_z);
        defer file.close();

        // Magic: "IVF1"
        var magic: [4]u8 = undefined;
        _ = try file.read(&magic);
        if (!std.mem.eql(u8, &magic, "IVF1")) {
            std.debug.print("index invalido: magic != IVF1\n", .{});
            return error.InvalidIndex;
        }

        const n = try file.readInt(u32);
        const k = try file.readInt(u32);
        const d = try file.readInt(u32);

        if (k != IVF_CLUSTERS or d != DIM) {
            std.debug.print("index incompativel: K={} D={}\n", .{ k, d });
            return error.InvalidIndex;
        }

        // Transposed centroids: DIM * IVF_CLUSTERS floats, align(32)
        const centroids_len = DIM * IVF_CLUSTERS;
        const centroids_t = try allocator.alignedAlloc(f32, 32, centroids_len);
        errdefer allocator.free(centroids_t);
        const centroids_bytes = std.mem.sliceAsBytes(centroids_t);
        _ = try file.read(centroids_bytes);

        // Block offsets: K+1 uint32
        var block_offsets: [IVF_CLUSTERS + 1]u32 = undefined;
        const offsets_bytes = std.mem.sliceAsBytes(block_offsets[0..]);
        _ = try file.read(offsets_bytes);

        const total_blocks: usize = @intCast(block_offsets[IVF_CLUSTERS]);
        const padded_n = total_blocks * 8;

        // Labels: padded_n uint8
        const labels = try allocator.alignedAlloc(u8, 64, padded_n);
        errdefer allocator.free(labels);
        _ = try file.read(labels);

        // Blocks: total_blocks * BLOCK_STRIDE int16, align(32)
        const blocks_len = total_blocks * BLOCK_STRIDE;
        const blocks = try allocator.alignedAlloc(i16, 32, blocks_len);
        errdefer allocator.free(blocks);
        const blocks_bytes = std.mem.sliceAsBytes(blocks);
        _ = try file.read(blocks_bytes);

        const mb = @as(f64, @floatFromInt(blocks_len * @sizeOf(i16) + padded_n + centroids_len * @sizeOf(f32) + (IVF_CLUSTERS + 1) * @sizeOf(u32))) / (1024.0 * 1024.0);
        std.debug.print("index IVF1 carregado (Zig): N={} K={} blocks={} memoria={d:.2} MB\n", .{ n, k, total_blocks, mb });

        return Dataset{
            .n = @intCast(n),
            .total_blocks = total_blocks,
            .padded_n = padded_n,
            .centroids_t = centroids_t,
            .block_offsets = block_offsets,
            .labels = labels,
            .blocks = blocks,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *Dataset) void {
        self.allocator.free(self.centroids_t);
        self.allocator.free(self.labels);
        self.allocator.free(self.blocks);
    }

    pub fn clusterBlockRange(self: *const Dataset, cluster: usize) struct { start: usize, end: usize } {
        return .{
            .start = @intCast(self.block_offsets[cluster]),
            .end = @intCast(self.block_offsets[cluster + 1]),
        };
    }
};
